### ISSUE TITLE
.NET 9.0 support strategy

### ISSUE THREAD

**Issue by sebastienros** [OrchardCMS/OrchardCore#16985]
.NET 9.0 should be GA on 11/12/2024 (next Tuesday). How do we want to proceed with the merging on the branch?

My suggestion is to
- Update the version and merge to main as soon as it is released?
- Ship 2.1 (we are currently at 2.0.2) since it contains important changes (including net9.0 support) though not breaking? (we are still supporting net8.0)
- Building the solution will require dotnet 9.0 sdk though. Since it's enforced by `global.json` it will also be a requirement even if you want to build/deploy for net8.0.

An alternative could be to release a `2.1` with the current `main` branch, and right after a `2.2` so users can still get the current fixed without having to deal with net9.0.

**Comment by hishamco** [comment]
IMHO no need to rush, so we should focus on quality, .NET 9.0 is a major release we need proper time for testing. I suggest merging 9.0 on the main then it will be a part of `2.1` - if we can defer the release - otherwise we publish `2.1` then `2.2` is the one that targets both .NET 8 & 9

**Comment by Piedone** [comment]
We should double-target .NET 8 and .NET 9, just as we did with 6 and 7, and how we always target LTS and latest if they're not the same. Then we can release that as 2.1, since it won't be breaking (as a consumer, you can keep using .NET 8).

**Comment by MikeAlhayek** [comment]
> We should double-target .NET 8 and .NET 9, just as we did with 6 and 7, and how we always target LTS and latest if they're not the same. Then we can release that as 2.1, since it won't be breaking (as a consumer, you can keep using .NET 8).

Agreed.

**Comment by hishamco** [comment]
> We should double-target .NET 8 and .NET 9, just as we did with 6 and 7

Once again when we target 10 we will delete the previous one?!!

**Comment by Piedone** [comment]
Yes, since 10 will again be an LTS.

**Comment by hishamco** [comment]
Also, we could align OC major release with the LTS version

**Comment by MikeAlhayek** [comment]
I don’t think we should drop support for .NET 8 in favor of .NET 10 unless there are critical APIs in .NET 10 that we absolutely need. Otherwise, we should continue supporting the lowest version possible. Dropping .NET 8 support would effectively force everyone using Orchard Core to also drop support for it, which could create unnecessary friction.

I’m open to dropping support if there are compelling new APIs in .NET 10 that we must have—similar to how we dropped support for .NET 6 when we gained access to new APIs like ArgumentNullException.ThrowIfNull() and other features introduced in .NET 7 and beyond.

**Comment by sebastienros** [comment]
> Otherwise, we should continue supporting the lowest version possible.

It's a matter of how many versions we want to support. Two seems fine because that's the minimum we can do to always have one LTS.

Dropping the previous LTS is what we have done so far. Upgrading to the new LTS is usually not an issue for OC, since whatever they use that targets the previous LTS will also work with the newer (a library targeting net6.0 still works for net8.0).

So yes, when net10.0 ships next year this is what we'll support as the target TFM. Like what we did last year when net8.0 shipped.

**Comment by sebastienros** [comment]
> Also, we could align OC major release with the LTS version

No. We don't have to. It would just leave us with breaking changes only once a year, on specific dates. We don't have any reason to do so.

**Comment by MikeAlhayek** [comment]
> So yes, when net10.0 ships next year this is what we'll support as the target TFM. Like what we did last year when net8.0 shipped.

Last year, we dropped .NET 6 to take advantage of APIs available only in .NET 7+ and to avoid adding version-specific checks. .NET 8 reaches end of life on November 10, 2026, coinciding with the release of .NET 10, making it logical to discontinue .NET 8 support then, as it will no longer be supported. However, dropping a framework should be part of a major release, as it could introduce breaking changes. We might consider aligning **at least one major release** with each .NET LTS version.

**Comment by sebastienros** [comment]
> Last year, we dropped .NET 6 to take advantage of APIs available only in .NET 7+ and to avoid adding version-specific checks.

No. We dropped .NET6 for the reasons I explained before. And as a consequence we were also able to do the things you mention.

I agree that dropping a version is breaking. And that we should reflect it. But it shouldn't wait for LTS releases (how I read Hisham's comment, I prefer your version).

**Comment by MikeAlhayek** [comment]
.net 9 was merged.

**Comment by anthnyg12b** [comment]
> .net 9 was merged.

hi @MikeAlhayek  what iteration of release 2.1.x was it merged into? Can you send me the PR as I don't see it in the code. Can we upgrade to 2.1.3 with .net 9 or do you recommend only using 2.1.2 and 2.1.1 for .net 9. We are currently on .net 8

**Comment by MikeAlhayek** [comment]
You should be able to use .net9 with v2. However, OC .net 9 was officially added with v3 which is still preview and not production ready. 

Add .net 9 to your project and see if you encounter any issues

**Comment by anthnyg12b** [comment]
Good Afternoon @MikeAlhayek ,

I did try updating my project(version 2.1.2) from .net 8 to .Net 9 and I encountered issues with the nuget package OrchardCore.Application.Cms.Core.Targets(2.1.2).  
Particularly 
    -  Detected package downgrade: Microsoft.AspNetCore.Authentication.OpenIdConnect from 9.0.0-rc.2.24474.3 to 8.0.11. 
And 
     - OrchardCore.Application.Cms.Core.Targets 2.1.2 -> Microsoft.AspNetCore.Authentication.OpenIdConnect (>= 8.0.11)
     - OrchardCore.Application.Cms.Core.Targets 2.1.2 -> Microsoft.Identity.Web 3.3.1 -> Microsoft.Identity.Web.TokenAcquisition 3.3.1 -> Microsoft.AspNetCore.Authentication.OpenIdConnect (>= 9.0.0-rc.2.24474.3) 

My guess would be the orchardcore nuget package is referencing .net 8 while the project has been upgraded to .net 9?

**Comment by MikeAlhayek** [comment]
@sebastienros can't he use .net9 on his project since OC supports minimum of .net8 or does it need to explicitly support .net9 too?

@anthnyg12b you can always use the 3.0 previews which support .net8 and .net9.

**Comment by anthnyg12b-hub** [comment]
hi @MikeAlhayek, where can I download those from?

**Comment by MikeAlhayek** [comment]
https://docs.orchardcore.net/en/latest/getting-started/preview-package-source/

**Comment by anthnyg12b** [comment]
hi @sebastienros / @MikeAlhayek  wanted to follow up and see if support for .net 9 has been added to release 2.1.4 as per 

![Image](https://github.com/user-attachments/assets/ff30abba-bcc9-4475-b9d7-4e68f051985f) below it appears to be merged into main?

**Comment by anthnyg12b** [comment]
hi @sebastienro / @MikeAlhayek,

What's interesting is the  transitive dependencies are different in the CrestApp you referred me to. OrchardCore.Application.Cms.Core.Targets(2.1.4) in our app references earlier versions of the OpenIdConnect and Microsoft.Identity.Web. While the CrestView references the latest. This is the reason .net 9 is working for CrestView and not our app. I upgraded to OrchardCore.Application.Cms.Core.Targets(2.1.4) from the visual studio nuget package manager. Any idea why my  transitive dependencies are different??

![Image](https://github.com/user-attachments/assets/a6c8888a-725d-43fc-a59f-b2aef97f059e)

**Comment by lamualfa** [comment]
@MikeAlhayek how did u use .NET 9 in Orchard Core? I got an error when using `occms` template. Already follows the [Preview Package Source](https://docs.orchardcore.net/en/latest/getting-started/preview-package-source/).

**`<REDACTED>.csproj`**

```csproj
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>net9.0</TargetFramework>
    <AspNetCoreHostingModel>InProcess</AspNetCoreHostingModel>
    <Nullable>enable</Nullable>
    <ImplicitUsings>enable</ImplicitUsings>
  </PropertyGroup>

  <ItemGroup>
    <Folder Include="wwwroot\" />
    <Folder Include="Localization\" />
  </ItemGroup>

  <!-- Watcher include and excludes -->
  <ItemGroup>
    <Watch Include="**\*.cs"
      Exclude="Recipes\**;Assets\**;node_modules\**\*;**\*.js.map;obj\**\*;bin\**\*" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="OrchardCore.Logging.NLog" Version="2.1.6" />
    <PackageReference Include="OrchardCore.Application.Cms.Targets" Version="2.1.6" />
  </ItemGroup>

</Project>
```

**`dotnet run --project <REDACTED>`**

```bash
Using launch settings from <REDACTED>/Properties/launchSettings.json...
/home/laode/Repositories/<REDACTED>/<REDACTED>/<REDACTED>/<REDACTED>.csproj : error NU1605: 
  Warning As Error: Detected package downgrade: System.Text.Json from 9.0.0-rc.2.24473.5 to 8.0.5. Reference the package directly from the project to sel
  ect a different version. 
   <REDACTED> -> OrchardCore.Application.Cms.Targets 2.1.6 -> Microsoft.Identity.Web 3.3.1 -> System.Text.Json (>= 9.0.0-rc.2.24473.5) 
   <REDACTED> -> OrchardCore.Application.Cms.Targets 2.1.6 -> System.Text.Json (>= 8.0.5)
/home/laode/Repositories/<REDACTED>/<REDACTED>/<REDACTED>/<REDACTED>.csproj : error NU1605: 
  Warning As Error: Detected package downgrade: Microsoft.AspNetCore.Authentication.OpenIdConnect from 9.0.0-rc.2.24474.3 to 8.0.11. Reference the packag
  e directly from the project to select a different version. 
   <REDACTED> -> OrchardCore.Application.Cms.Targets 2.1.6 -> Microsoft.Identity.Web 3.3.1 -> Microsoft.Identity.Web.TokenAcquisition 3.3.1 -> Micros
  oft.AspNetCore.Authentication.OpenIdConnect (>= 9.0.0-rc.2.24474.3) 
   <REDACTED> -> OrchardCore.Application.Cms.Targets 2.1.6 -> Microsoft.AspNetCore.Authentication.OpenIdConnect (>= 8.0.11)
/home/laode/Repositories/<REDACTED>/<REDACTED>/<REDACTED>/<REDACTED>.csproj : error NU1605: 
  Warning As Error: Detected package downgrade: Microsoft.Extensions.Caching.Memory from 9.0.0-rc.2.24473.5 to 8.0.1. Reference the package directly from
   the project to select a different version. 
   <REDACTED> -> OrchardCore.Application.Cms.Targets 2.1.6 -> Microsoft.Identity.Web 3.3.1 -> Microsoft.Identity.Web.TokenCache 3.3.1 -> Microsoft.Ex
  tensions.Caching.Memory (>= 9.0.0-rc.2.24473.5) 
   <REDACTED> -> OrchardCore.Application.Cms.Targets 2.1.6 -> Microsoft.Extensions.Caching.Memory (>= 8.0.1)

The build failed. Fix the build errors and run again.
```

**Comment by lamualfa** [comment]
> You may have a non-OC packages that needs to also target a higher version. For example System.Text.Json in your case.
> 
> For reference, the following project targets both .net8 and .net9 at the same time
> 
> https://github.com/CrestApps/CrestApps.OrchardCore

It's freshly generated using `dotnet new occms`. So no added packages yet.

Sorry, I'm new in .NET ecosystem.  I guess .NET has the same package system with Python where all packages are globally installed by default. I'll figure it out. Thanks.

**Comment by SamLeurink** [comment]
I'm sorry for commenting on a thread that has been stale for nearly half a year, but I think it is the best place to do so.

I'm getting the same errors as @lamualfa where package downgrades are detected on net9.0. From my understanding this is because there is no package with TargetFramework net9.0 on nuget.org. However I do see in this thread (and the release_ci.yml file) that support for .net9 should already be available. 

Is there a reason why an old version of the release pipeline is used for releasing new versions? And if so, when should we expect OrchardCore to have net9.0 as their TargetFramework? @MikeAlhayek

**Comment by MikeAlhayek** [comment]
@SamLeurink if you are using version 2.1 or 2.2, both support .net8 and .net9.

If you are using the previews or the main branch, they you need .net9.

If it does not work, read through the error, you may have a bad package in your project. Or you may need to clean up the solution or delete the .vs folder if you just switched framework targets.

**Comment by SamLeurink** [comment]
@MikeAlhayek the thing is that while .net9 is a compatible framework, it isn't a TargetFramework( see https://www.nuget.org/packages/OrchardCore/2.2.0#supportedframeworks-body-tab and https://nuget.info/packages/OrchardCore/2.2.0), which means that when packaging the PackageReference versions of included packages, that are decided in Directory.Packages.props, will look at the versions of .net8 instead of the intended .net9.

This causes exceptions like such: 
`error NU1605:  MyProject -> OrchardCore.Application.Cms.Targets 2.2.0 -> OpenIddict.Validation.SystemNetHttp 7.0.0 -> Microsoft.Extensions.Http.Resilience (>= 9.6.0) 
error NU1605:  MyProject -> OrchardCore.Application.Cms.Targets 2.2.0 -> Microsoft.Extensions.Http.Resilience (>= 8.10.0)`
Which only occurs in referenced packages that have both .net8 and .net9 as TF and have different PackageReference versions based on TF( see https://www.nuget.org/packages/OpenIddict.Validation.SystemNetHttp#supportedframeworks-body-tab and https://github.com/openiddict/openiddict-core/blob/dev/Directory.Packages.props)

The previews and the main branch are all set for releasing .net9, which is why on these branches there will probably be no issue, but on the version branches the pipeline is as follows: https://github.com/OrchardCMS/OrchardCore/actions/runs/17077747541/workflow.

Is this by choice? Cause if it is I can try to have a look at using the previews but I would prefer to keep using stable releases.

**Comment by MikeAlhayek** [comment]
The package support .net8 and .net9. If you recently upgraded your project from .net8 to .net9, you may need to run clean and manually delete the .vs folder. 

It your project targets .net8 or .net9 it will work because the same dll works with both frameworks.

My AI modules https://github.com/CrestApps/CrestApps.OrchardCore/tree/release/1.2 is using nuget packages with .net8 and .net9. 

You may need to add conditions to change the package version depending on the version you want like I did here: https://github.com/CrestApps/CrestApps.OrchardCore/blob/release/1.2/Directory.Packages.props#L93-L102

**Comment by SamLeurink** [comment]
I would prefer not to add package versions for packages that aren't directly referenced by my project. It seems both unnecessary and confusing for developers after me. I'm fine with putting them there for now but would like to know when I will be able to remove them again.

Therefore my question still remains: When will .net9 be a TF for the OC releases?

**Comment by MikeAlhayek** [comment]
Yes. Do not add packages unless you need them. I am saying that maybe you are running into an issue where your project depends on packages that have different version per framework and you'll have to do what I did.

OrchardCore already support .net8 and .net9 but we do not ship a framework specific version. So OC 2.0 will work with .net8 and .net9.  OC 3 "when shipped will only be supporting .net10+ if it's available by then"

**Comment by SamLeurink** [comment]
While there may not be a framework specific version, the pipeline is being built using .net8 which in turn means that referenced packages that have .net specific versioning will chose the version designated for .net8, creating an opportunity for version mismatches when referencing OC in .net9. 

I'll go back to the example I gave earlier of `Microsoft.Extensions.Http.Resilience` and `OpenIddict.Validation.SystemNetHttp`. 

In Directory.Packages.Props, the version for Resillience is set to 8.10 (https://github.com/OrchardCMS/OrchardCore/blob/dff988725fbfae55091cebf7bb4f38a99a006de5/Directory.Packages.props#L81), which is fine by SystemNetHttp when compiling for .net8(https://github.com/openiddict/openiddict-core/blob/dev/Directory.Packages.props#L120) but once it is referenced with .net9 as TF the version of Resillience referenced by SystemNetHttp is 9.6.0, causing a mismatch. 

Any project that references OC from .net9 will have this issue, unless there is specific version control in their own project setting SystemNetHttp to 9.6.0.

As I see it there are 3 things that can be done to solve this:

1. Don't solve it, accept that users of .net 9 will have to set versions of a few packages referenced by OC in their own project. A note for new users that want to use OC with .net9 would be nice however.
2. Make OC use both .net8 and .net9 as their TF. This would require some changes, both in the pipeline and in Directory.Packages.Props. Even though most of these changes are already implemented on the main branch, there will probably be a need to make sure everything stays working as intended. So there is probably some regression testing that needs to be done.
3. Upgrade the packages in question. In the case of `Microsoft.Extensions.Http.Resilience` an upgrade to 9.6.0 (or even 9.8.0) would solve the issue for .net9 without influencing .net8 as they are both listed as TF in that version(https://www.nuget.org/packages/Microsoft.Extensions.Http.Resilience/9.6.0#supportedframeworks-body-tab). I'll be able to help in identifying which packages would need an upgrade. However the same regression tests as previous option still need to be done.

Hopefully I'm not missing anything or sayig things that are too outlandish. Would love to here your input on this!

**Comment by MikeAlhayek** [comment]
If your project is targeting multiple frameworks, then you need to do something like I did here https://github.com/CrestApps/CrestApps.OrchardCore/blob/release/1.2/Directory.Packages.props#L93-L102 because these Microsoft packages are a version for each framework. You can't use the 9.0 version with 
Net8. 

If your project onto targets .net9, then just use the latest package and you don't need to worry about this complexity and this will allow you to use OC "as long as you use Microsoft .net9 packages"

**Comment by sebastienros** [comment]
> You can't use the 9.0 version with Net8.

@MikeAlhayek We can with the recent versions, this has been fixed. It's case by case though but for `Microsoft.Extensions.Http.Resilience` for instance this is fine now, the version targeting net8.0 only references net8.0 runtime ones.

> an upgrade to 9.6.0 (or even 9.8.0) would solve the issue for .net9 without influencing .net8 as they are both listed as TF in that version

@SamLeurink This is not fully correct. They were already doing so before AFAIK but transitively referencing net9.0 runtime ones. This is why we need to check each version/release to ensure this is not broken when we check to upgrade.

Otherwise we can also target different package versions based on the tfm if this adds some value (rare case though)
